processing/src/main/java/io/druid/segment/DimensionMerger.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import com.google.common.io.OutputSupplier;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.Buffer;
+import java.nio.IntBuffer;
+import java.util.List;
+
+/**
+ * Processing related interface
+ *
+ * A DimensionMerger is a per-dimension stateful object that encapsulates type-specific operations and data structures
+ * used during the segment merging process (i.e., work done by IndexMerger/IndexMergerV9).
+ *
+ * This object is responsible for:
+ *   - merging any relevant structures from the segments (e.g., encoding dictionaries)
+ *   - writing the merged column data and any merged indexing structures (e.g., dictionaries, bitmaps) to disk
+ *
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ */
+public interface DimensionMerger<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
+{
+  /**
+   * Initialize any writer objects that can be instantiated before data from the segments has been processed/merged.
+   * @throws IOException
+   */
+  public void initWriters() throws IOException;
+
+
+  /**
+   * Read the segments to be merged and combine their indexes/data structures internally.
+   * (e.g., merging of dictionaries for String dimensions).
+   *
+   * @param adapters List of adapters to be merged.
+   * @throws IOException
+   */
+  public void mergeAcrossSegments(List<IndexableAdapter> adapters) throws IOException;
```

Comment:
addColumnValue wouldn't be called within mergeAcrossSegments, it is called later when iterating through the merged Rowboat iterable

"mergeSegmentIndices" might be a better name than "mergeAcrossSegments"...